File: Longest Substring Without Repeating Characters.cpp
int lengthOfLongestSubstring(string s)
{
    int maxLen = 0;
    int start = 0;
    vector<int> positions(256, -1);
    for (int i = 0; i < s.size(); i++)
    {
        if (positions[s[i]] != -1)
        {
            maxLen = max(maxLen, i - start);
            int j = start;
            for (; j <= positions[s[i]]; j++)
            {
                positions[s[j]] = -1;
            }
            start = j;
        }

        positions[s[i]] = i;
    }

    return max(maxLen, int(s.size()) - start);
}

File: Leetcode/Longest Substring Without Repeating Characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        maxLen = 0
        start = 0
        positions = {}
        for i in range(len(s)):
            if s[i] in positions and positions[s[i]] != -1:
                maxLen = max(maxLen, i - start)
                j = start
                while j <= positions[s[i]]:
                    positions[s[j]] = -1
                    j += 1
                start = j
            positions[s[i]] = i

        return max(maxLen, len(s) - start)
